File: leetcode/cpp/plusOne.cpp

// leetcode.com  66. Plus One

class Solution
{
  public:
    vector<int> plusOne(vector<int> &digits)
    {
        if (digits.size() == 0)
            return digits;

        int carry = 1, sum = 0;

        for (int i = digits.size() - 1; i >= 0; --i)
        {
            if (carry == 0)
                return digits;

            sum = digits[i] + carry;
            carry = sum / 10;
            digits[i] = sum % 10;
        }

        if (carry)
            digits.insert(digits.begin(), carry);

        return digits;
    }
};

File: leetcode/java/src/main/java/com/vkeonline/leetcode/p000/PlusOne.java
package com.vkeonline.leetcode.p000;

/**
 * Leetcode [E]: 66. Plus One
 * @author csgear
 */
public class PlusOne {
    public int[] plusOne(int[] digits) {
        int len = digits.length, carry = 1;

        while (--len >= 0 && carry != 0) {
            carry = digits[len] + 1 >= 10 ? 1 : 0;
            digits[len] = (digits[len] + 1) % 10;
        }

        int[] ans = new int[digits.length + 1];

        // if carry happens at the very first digits, set the first to be 1, like 999
        if (carry == 1) {
            ans[0] = 1;
        }

        return carry == 1 ? ans : digits;
    }
}
